docs/howto/use_modules_and_layers.md

```diff
@@ -0,0 +1,270 @@
+def createUser(user: User): ZIO[DBConnection, Nothing, Unit] = UIO(???)
+
+val user: User = User(UserId(1234), "Chet")
+val created: ZIO[DBConnection, Nothing, Boolean] = for {
+  maybeUser <- getUser(user.id)
+  res       <- maybeUser.fold(createUser(user).as(true))( _ => ZIO.succeed(false))
+} yield res
+```
+
+To run the program we must supply a `DBConnection` through `provide`, before feeding it to ZIO runtime.
+
+```scala
+val dbConnection: DBConnection = ???
+val runnable: ZIO[Any, Nothing, boolean] = created.provide(dbConnection)
+
+val finallyCreated  = runtime.unsafeRun(runnable)
+```
+
+Notice that the act of `provide`ing an effect with its environment eliminates the environment dependency in the resulting effect type, well represented by type `Any` of the resulting environment.
+
+In general we need more than just a DB connection though. We need components that enable us to perform different operations, and we need to be able to wire them together. This is what _modules_ are for.
+
+## Our first ZIO module
+We will see now how to define modules and use them to create different application layers relying on each other. The core idea is that a layer depends on the layers imediately below but it is completely agnostic about their internal implementation.
+ 
+This formulation of module pattern is _the way_ ZIO manages dependencies between application components, giving extreme power in terms of compositionality and offering the capability to swap implementations. This is particularly useful during the testing/mocking phase. 
+
+### What is a module?
+A module is a group of functions that deals with only one concern. Keeping the scope of a module limited improves our ability to understand code, in that we need to focus
+ only on one topic at a time without juggling with too many concepts together in our head.
+ 
+`ZIO` iself provides the basic capabilities through modules, e.g. see how `ZEnv` is defined.
+
+### The module recipe
+Let's build a module for user data access, following these simple steps:
+
+1. Define an object that gives the name to the module, this can be (not necessarily) a package object
+1. Within the module object define a `trait Service` that defines the interface our module is exposing, in our case 2 methods to retrieve a product details, and one to retrieve the related products
+1. Within the module object define a type alias like `type ModuleName = Has[Service]` (see below for details on `Has`)
+1. Within the module object define the different implementations of `ModuleName` through `ZLayer` (see below for details on `ZLayer`)
+
+```scala mdoc:silent
+import zio.{Has, ZLayer}
+
+type UserRepo = Has[UserRepo.Service]
+
+object UserRepo {
+  trait Service {
+    def getProductDetails(productId: ProductId): IO[DBError, Product]
+    def getRelatedProducts(productId: ProductId): IO[DBError, List[Product]]
+  }
+
+  val testRepo: ZLayer[Any, Nothing, UserRepo] = ZLayer.succeed(???)
+}
+```
+
+```scala mdoc:reset:silent
```

Comment:
I think you should already have these imports in scope from above so you don't need to import them again.